### ISSUE TITLE
Make encode-alist resilient against empty values

### ISSUE THREAD

**Issue by equill** [equill/neo4cl#20]
This _should_ be as simple as dropping in `remove-if` with a lambda to test for null `car` values,  but naturally it also needs tests to make sure it does what's expected.

Discovered while investigating #19

**Comment by equill** [comment]
Relates to #26

**Comment by mmontone** [comment]
What is the issue with the null cdr? Isn't it just encoded as a Neo4j `null` ?

**Comment by mmontone** [comment]
This prevents from encoding a dictionary with a member with a null value, that could appear in practice (should be valid).

**Comment by equill** [comment]
On re-reading the spec, I think you're right: it doesn't explicitly say _what_ might qualify as valid values.

**Comment by mmontone** [comment]
My thinking is: the intention is encode some association lists as dictionaries, so anything encodable should be allowed on the value side (cdr), and any valid keys should be allowed on the key side (cdr).
Perhaps it would be good to rename to something like `dictionary-encodable-alist-p` predicate to make that clear.

**Comment by mmontone** [comment]
Yes, but `("foo" ("bar" "baz") "quux")` won't be caught as an a-list, because "foo" and "quux" are not conses, but `(("foo") ("bar" . "baz") ("quux"))` would, but that's a valid encodable alist.

**Comment by mmontone** [comment]
Something like this perhaps:

```lisp
(defun neo4cl::dictionary-encodable-alist-p (lst)
  "Check whether the argument is an association list that can be encoded as a dictionary."
  (and (listp lst)
       (or (null lst)
           (every #'(lambda (pair)
                      (and (consp pair)
                           (stringp (car pair))))
                  lst))))
```

Not sure if anything other than strings could be encoded as dictionary keys.

**Comment by mmontone** [comment]
> I'd complain that this is an unwieldy function name, but this is Common Lisp :)

Yes, but you are not trying to determine if a list is an association list or not, you are trying to determine if an association list can be encoded as a neo4j dictionary or not :)

**Comment by mmontone** [comment]
![image](https://user-images.githubusercontent.com/436110/225368288-3f7bd8b3-14c1-4a51-83fb-475c1bfe2cb7.png)

**Comment by equill** [comment]
I hate you.
What's worse is that you're right about the keys: https://neo4j.com/docs/bolt/current/packstream/#data-type-dictionary says "keys must be a String". The only thing it says about values is that they may be "a mixture of types" - it doesn't specify any that are or are not permitted implying that `Null` is valid.

OK, fine. I do it your way.

**Comment by equill** [comment]
At this rate, you're going to get credit as a co-maintainer.

**Comment by equill** [comment]
Commited, and now we have a `development` branch for accumulating changes until it's time to declare a new version. This commit is on that branch.
